ignite/metrics/psnr.py

```diff
@@ -0,0 +1,118 @@
+            process_function’s output into the form expected by the metric.
+        device (str or torch.device): specifies which device updates are accumulated on.
+            Setting the metric’s device to be the same as your update arguments ensures
+            the update method is non-blocking. By default, CPU.
+
+    Example:
+
+    To use with ``Engine`` and ``process_function``, simply attach the metric instance to the engine.
+    The output of the engine's ``process_function`` needs to be in format of
+    ``(y_pred, y)`` or ``{'y_pred': y_pred, 'y': y, ...}``.
+
+    .. code-block:: python
+
+        def process_function(engine, batch):
+            # ...
+            return y_pred, y
+        engine = Engine(process_function)
+        psnr = PSNR()
+        psnr.attach(engine, "psnr")
+        # ...
+        state = engine.run(data)
+        print(f"PSNR: {state.metrics['psnr']}")
+
+    .. versionadded:: 0.5.0
+    """
+
+    def __init__(
+        self,
+        data_range: Optional[Union[int, float]] = None,
+        output_transform: Callable = lambda x: x,
+        device: Union[str, torch.device] = torch.device("cpu"),
+    ):
+        super().__init__(output_transform=output_transform, device=device)
+        self.data_range = data_range
+
+    def _check_shape_dtype(self, output: Sequence[torch.Tensor]) -> None:
+        y_pred, y = output
+        if y_pred.dtype != y.dtype:
+            raise TypeError(
+                f"Expected y_pred and y to have the same data type. Got y_pred: {y_pred.dtype} and y: {y.dtype}."
+            )
+
+        if y_pred.shape != y.shape:
+            raise ValueError(
+                f"Expected y_pred and y to have the same shape. Got y_pred: {y_pred.shape} and y: {y.shape}."
+            )
+
+    @reinit__is_reduced
+    def reset(self) -> None:
+        self._sum_of_batchwise_psnr = 0  # type: Union[int, torch.Tensor]
+        self._num_examples = 0
+
+    @reinit__is_reduced
+    def update(self, output: Sequence[torch.Tensor]) -> None:
+        self._check_shape_dtype(output)
+        y_pred, y = output[0].detach(), output[1].detach()
+        data_range = self.data_range
+
+        if data_range is None:
+            dmin, dmax = _dtype_range[y.dtype]
+            true_min, true_max = y.min(), y.max()
+            if true_max > dmax or true_min < dmin:
+                raise ValueError(
+                    "y has intensity values outside the range expected "
+                    "for its data type. Please manually specify the data_range."
+                )
+            if true_min >= 0:
+                # most common case (255 for uint8, 1 for float)
+                data_range = dmax
+            else:
+                data_range = dmax - dmin
+
+        rmse_error = torch.pow(y_pred.double() - y.view_as(y_pred).double(), 2).mean(dim=list(range(1, y.ndim))).sqrt()
+        self._sum_of_batchwise_psnr += 20.0 * torch.log10(data_range / rmse_error).to(self._device)
```

Comment:
Should we call `torch.sum()` here or in the `compute()`?
I saw the other metrics calling `torch.sum()` here. How about the memory consumption?

Reply:
Sorry, i'll bother you again with if we are certain now that this implementation is correct one now :) ?

Reply:
I think I am certain about it.
Right now I am computing the MSE (or RMSE) per image of a batch and psnr from that, resulting in psnr shape of [batch_size] with a requirement of dim 1 being batch. The main issue is I am computing MSE per batch, not per image at the start.

Do you have any other way in mind?

Reply:
> The main issue is I am computing MSE per batch, not per image at the start.

which part of the code correspond to that (at start) ?
Anyway, let me take sometime to check it from my side and we'll see.

Reply:
This one
```py
mse_error = torch.pow(y_pred.double() - y.view_as(y_pred).double(), 2).mean()
```
That was around the start commit of the PR

Reply:
ah, OK I see what you wanted to say.

Reply:
About your question on where to call the sum. Let's call it in update as in other metrics. Such that we could also remove `# type: ignore[arg-type]` from compute.


Another point, IMO, it is faster to compute square of data_range than rmse. Let's do 10 * log10(max / mse).